fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/impl/datagen/SoundTypeBuilderImpl.java

```diff
@@ -0,0 +1,217 @@
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.datagen.v1.builder.SoundTypeBuilder;
+
+public final class SoundTypeBuilderImpl implements SoundTypeBuilder {
+	private static final Logger LOGGER = LoggerFactory.getLogger(FabricDataGenHelper.class);
+
+	private SoundCategory category = SoundCategory.NEUTRAL;
+	@Nullable
+	private String subtitle;
+	private final List<Entry> sounds = new ArrayList<>();
+
+	public SoundTypeBuilderImpl() { }
+
+	@Override
+	public SoundTypeBuilder category(SoundCategory category) {
+		Preconditions.checkArgument(category != null, "Sound event category must not be null.");
+		this.category = category;
+		return this;
+	}
+
+	@Override
+	public SoundTypeBuilder subtitle(@Nullable String subtitle) {
+		this.subtitle = subtitle;
+		return this;
+	}
+
+	@Override
+	public SoundTypeBuilder sound(EntryBuilder sound) {
+		Preconditions.checkArgument(sound != null, "Sound must not be null.");
+		sounds.add(((EntryBuilderImpl) sound).build(""));
+		return this;
+	}
+
+	@Override
+	public SoundTypeBuilder sound(EntryBuilder sound, int count) {
+		Preconditions.checkArgument(sound != null, "Sound must not be null.");
+		Preconditions.checkArgument(count > 0, "Count must be greater than zero.");
+
+		for (int i = 1; i <= count; i++) {
+			sounds.add(((EntryBuilderImpl) sound).build("" + i));
+		}
+
+		return this;
+	}
+
+	public SoundType build() {
+		Preconditions.checkState(!sounds.isEmpty(), "Sound definition must have at least one sound file");
+
+		for (Entry sound : sounds) {
+			if (sound.type() == RegistrationType.SOUND_EVENT) {
+				Registries.SOUND_EVENT.getOptionalValue(sound.name()).orElseThrow(() -> new IllegalStateException("References sound event " + sound.name() + " does not exist"));
+			}
+		}
+
+		return new SoundType(sounds, category, Optional.ofNullable(subtitle));
+	}
+
+	public record SoundType(List<Entry> sounds, SoundCategory category, Optional<String> subtitle) {
+		private static final Map<String, SoundCategory> CATEGORIES = Arrays.stream(SoundCategory.values()).collect(Collectors.toMap(SoundCategory::getName, Function.identity()));
+		private static final Codec<SoundCategory> SOUND_CATEGORY_CODEC = Codec.stringResolver(SoundCategory::getName, name -> CATEGORIES.getOrDefault(name.toLowerCase(Locale.ROOT), SoundCategory.NEUTRAL));
+		public static final Codec<SoundType> CODEC = RecordCodecBuilder.create(i -> i.group(
+				Entry.CODEC.listOf().fieldOf("sounds").forGetter(SoundType::sounds),
+				SOUND_CATEGORY_CODEC.fieldOf("category").forGetter(SoundType::category),
+				Codec.STRING.optionalFieldOf("subtitle").forGetter(SoundType::subtitle)
+		).apply(i, SoundType::new));
+	}
+
+	private record Entry(Identifier name, RegistrationType type, float volume, float pitch, int weight, int attenuationDistance, boolean stream, boolean preload) {
+		private static final Codec<Entry> MAP_CODEC = RecordCodecBuilder.create(i -> i.group(
```

Comment:
Pretty sure this exists in vanilla right? Or does it still deserialize with gson?

Reply:
As explained in the PR description, vanilla still deserializes with gson and there's no actual support for serializing. The vanilla classes are also not nice to work with. That's why the builders are here.